Dockerfile

```diff
@@ -34,3 +34,9 @@ RUN yum update -y \
   && rm stack.tgz \
   && mv stack /usr/local/bin/ \
   && cd .. && rm -r /tmp/stack
+
+RUN useradd -m docker
+RUN mkdir -p /.stack
+RUN chown -R docker:docker /.stack
+
+USER docker
```

Comment:
Two comments here:

- What's the point of the `/.stack` directory? This image doesn't do anything special with it. If any downstream images need it (for example if they set `STACK_ROOT=/.stack`), then they'll have to create it. 
- Can we call the user something other than `docker`? I'd prefer `stack` or `haskell`. Also you could make it a build time argument to avoid repeating it: `ARG USER=stack; RUN useradd -m "$USER"; USER $USER`.

Reply:
> If any downstream images need it (for example if they set STACK_ROOT=/.stack), then they'll have to create it.

You can't create volumes as a non-root user. You have to create the directory before-hand, assign permissions, then they get inherited.

Reply:
> You can't create volumes as a non-root user.

Huh?

``` sh
$ docker build --tag ex20210406 . && 
  docker run --interactive --rm --tty --volume "$(pwd):/work" --workdir /work ex20210406 \
  sh -o xtrace -c 'pwd && ls -l && cat Dockerfile'
Sending build context to Docker daemon  2.048kB
Step 1/4 : FROM ubuntu:20.04
 ---> f643c72bc252
Step 2/4 : ARG USER=someone
 ---> Using cache
 ---> 3208ee85ea5b
Step 3/4 : RUN useradd --create-home "$USER"
 ---> Using cache
 ---> 36c24face137
Step 4/4 : USER "$USER"
 ---> Using cache
 ---> c599e9317d21
Successfully built c599e9317d21
Successfully tagged ex20210406:latest
+ pwd
/work
+ ls -l
total 4
-rw-rw-r-- 1 someone someone 82 Apr  6 18:44 Dockerfile
+ cat Dockerfile
FROM ubuntu:20.04
ARG USER=someone
RUN useradd --create-home "$USER"
USER "$USER"
```

Reply:
Sorry, I mean docker-compose will mess up those permissions. See https://github.com/docker/compose/issues/5507

Reply:
I wasn't able to reproduce the problem with Docker Compose either:

``` sh
/t/tmp.ER4g9jofa4 » cat Dockerfile
FROM ubuntu:20.04
ARG USER=someone
RUN useradd --create-home "$USER"
USER "$USER"

/t/tmp.ER4g9jofa4 » cat docker-compose.yaml
version: '3.7'
services:
  example:
    build: .
    volumes:
      - $PWD:/work
    working_dir: /work

/t/tmp.ER4g9jofa4 » docker-compose run --rm example sh -o xtrace -c 'ls -ld /work'
+ ls -ld /work
drwx------ 2 someone someone 4096 Apr  6 19:12 /work
```

But apparently the issue is talking about folders that don't already exist on the host, so let me try that out real quick.

Reply:
That does indeed show the problem:

``` sh
/t/tmp.ER4g9jofa4 » cat docker-compose.yaml
version: '3.7'
services:
  example:
    build: .
    volumes:
      - does-not-exist:/does-not-exist
volumes:
  does-not-exist: null

/t/tmp.ER4g9jofa4 » docker-compose run --rm example sh -o xtrace -c 'ls -ld /does-not-exist'
Creating volume "tmper4g9jofa4_does-not-exist" with default driver
+ ls -ld /does-not-exist
drwxr-xr-x 2 root root 4096 Apr  6 19:16 /does-not-exist
```

Reply:
And the final piece, showing that a directory already made in the container has the correct permissions even when used as a volume:

``` sh
/t/tmp.ER4g9jofa4 » cat Dockerfile 
FROM ubuntu:20.04
ARG USER=someone
RUN useradd --create-home "$USER"
USER "$USER"
RUN mkdir "$HOME/already-exists"

/t/tmp.ER4g9jofa4 » cat docker-compose.yaml 
version: '3.7'
services:
  example:
    build: .
    volumes:
      - already-exists:/home/someone/already-exists
volumes:
  already-exists: null

/t/tmp.ER4g9jofa4 » docker-compose run --rm example sh -o xtrace -c 'ls -ld $HOME/already-exists'
+ ls -ld /home/someone/already-exists
drwxr-xr-x 2 someone someone 4096 Apr  6 19:19 /home/someone/already-exists
```

Reply:
Bringing it back around to my original remark, it's up to any downstream containers to create directories/volumes that they're interested in. I don't think that this image needs to bother creating the `/.stack` directory.

Reply:
The problem is if anything creates it before the downstream container it has the wrong permissions.

IIRC for us that something was `stack install`.

Reply:
That may be true, but what I'm saying is that this image doesn't care about `/.stack` at all. With these changes you're preemptively making a directory because some other container may want that directory. There's no need -- the other container can make the directory itself.

Reply:
I tried to create `/work` in docker-stack-develop instead and got a permission denied error.

https://github.com/EdutainmentLIVE/docker-stack-develop/pull/27/commits/4d7f785a3ed3799f0361be76c5bc36b8a9df2dd2

Reply:
Nevermind, didn't realize user wasn't root at that point in time and made it explicit. It's getting to the `stack install` part I remember failing... but we'll see what happens.